src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,181 @@
+package com.thealgorithms.strings;
+
+import java.util.*;
+
+/**
+ * Aho-Corasick String Matching Algorithm Implementation
+ *
+ * This code implements the Aho-Corasick algorithm, which is used for efficient
+ * string matching in a given text. It can find multiple patterns simultaneously
+ * and records their positions in the text.
+ *
+ * Author: Prabhat-Kumar-42
+ * GitHub: https://github.com/Prabhat-Kumar-42
+ */
+public class AhoCorasick {
+
+    // Trie Node Class
+    class Node {
+        HashMap<Character, Node> child = new HashMap<>();
+        Node suffix_link;
+        Node output_link;
+        int pattern_ind;
+
+        Node() {
+            this.suffix_link = null;
+            this.output_link = null;
+            this.pattern_ind = -1;
+        }
+    }
+
+    Node root = null; // The root node of the Aho-Corasick trie
+    private ArrayList<ArrayList<Integer>> res; // Stores the positions where patterns are found in the text

```

Comment:
What do you think about storing the result in another data structure, for example:
```suggestion
    private HashMap<String, ArrayList<Integer>> res;
```
so res['somePattern'] would be a list of positions where `somePattern` occurs.

Reply:
> What do you think about storing the result in another data structure, for example:
> 
> so res['somePattern'] would be a list of positions where `somePattern` occurs.

Hey @vil02 ,

I can do that, but it will cost more space and time.

The current algorithm uses indices to calculate positions, as it is more efficient in terms of calculation and comparison within the algorithm. If I implement it using strings, the trie data structure will need space according to the length of the string that needs to be saved in it. Additionally, I would still need to find a way to calculate the position of words, which in the above case can be solved using the stored integer.

**TL;DR - 1**: I can do it, but it will cost both time and space.

However, what I can do to somewhat mitigate this is to add a new `Hash<String, ArrayList<Integer>>` called `positionByString`. This data structure will use the stored data in `res` and the provided dictionary or set of patterns (in our case, the pattern array provided in the code) to store the combined result. This will keep the time complexity the same and will only require extra time to process `res` and the pattern array to populate `positionByString`.

**TL;DR - 2:** While still keeping `res`, can create a `Hash<String, ArrayList<Integer>>` which utilizes data from `res` and the `pattern array` to mitigate the overhead.

I would greatly appreciate hearing your thoughts on this.

Please feel free to share your insights and opinions.

Reply:
> What do you think about storing the result in another data structure, for example:
> 
> so res['somePattern'] would be a list of positions where `somePattern` occurs.

I ended up going with the **TL;DR - 2**, I have implemented the added functionality along with the test cases.

I would say it is done, but I would appreciate your feedback regarding this.

If you feel like any changes should be made, please do tell me.